metricbeat/helper/prometheus/prometheus.go

```diff
@@ -76,110 +86,224 @@ func (p *prometheus) GetFamilies() ([]*dto.MetricFamily, error) {
 	}
 
 	families := []*dto.MetricFamily{}
-	for {
-		mf := &dto.MetricFamily{}
-		err = decoder.Decode(mf)
-		if err != nil {
-			if err == io.EOF {
-				break
+	var wg sync.WaitGroup
+	var decoderMutex = &sync.Mutex{}
+	var familiesMutex = &sync.Mutex{}
+	var done = false
+	var count = 0
+	for !done && count < maxGetFamiliesSubFuns {
+		wg.Add(1)
+		count++
+		go func(id int) {
+			for !done {
+				mf := &dto.MetricFamily{}
+				decoderMutex.Lock()
+				err = decoder.Decode(mf)
+				decoderMutex.Unlock()
+				if err != nil {
+					if err == io.EOF {
+						done = true
+					}
+				} else {
+					if familyPrefix == nil {
+						familiesMutex.Lock()
+						families = append(families, mf)
+						familiesMutex.Unlock()
+					} else {
+						found := false
+						for _, prefix := range familyPrefix {
+							if strings.HasPrefix(mf.GetName(), prefix) {
+								familiesMutex.Lock()
+								families = append(families, mf)
+								familiesMutex.Unlock()
+								found = true
+								break
+							}
+						}
+						if !found {
+							mf.Reset()
+						}
+					}
+				}
 			}
-		} else {
-			families = append(families, mf)
-		}
+			wg.Done()
+		}(count)
 	}
+	wg.Wait()
 
 	return families, nil
 }
 
+// KeyLabelKeys defines the primary and secondary key labels
+type KeyLabelKeys struct {
+	// Primary key label keys
+	Primary map[string]struct{}
+
+	// Secondary key label keys
+	Secondary map[string]struct{}
+}
+
 // MetricsMapping defines mapping settings for Prometheus metrics, to be used with `GetProcessedMetrics`
 type MetricsMapping struct {
+	// MetricSet Name -- for dbg
+	MetricSetName string
+
+	// Metrics Family name prefix
+	FamilyPrefix []string
+
 	// Metrics translates from from prometheus metric name to Metricbeat fields
 	Metrics map[string]MetricMap
 
+	// InfoMetrics translates from prometheus info metric name to Metricbeat fields
+	InfoMetrics map[string]MetricMap
+
 	// Labels translate from prometheus label names to Metricbeat fields
 	Labels map[string]LabelMap
 
+	// Primary and secondary key label keys (optional)
+	KeyLabels KeyLabelKeys
+
 	// ExtraFields adds the given fields to all events coming from `GetProcessedMetrics`
 	ExtraFields map[string]string
 }
 
-func (p *prometheus) GetProcessedMetrics(mapping *MetricsMapping) ([]common.MapStr, error) {
-	families, err := p.GetFamilies()
-	if err != nil {
-		return nil, err
-	}
-
-	eventsMap := map[string]common.MapStr{}
-	infoMetrics := []*infoMetricData{}
+func processMetrics(eventsMap *eventsMaps, mapping *MetricsMapping, families []*dto.MetricFamily) {
 	for _, family := range families {
-		for _, metric := range family.GetMetric() {
-			m, ok := mapping.Metrics[family.GetName()]
+		var wg sync.WaitGroup
 
-			// Ignore unknown metrics
-			if !ok {
-				continue
-			}
+		if _, ok := mapping.InfoMetrics[family.GetName()]; ok {
```

Comment:
what's going on here? if family is in the InfoMetrics array it means you already processed the family, so you're skipping it?

Reply:
No processMetrics processes non InfoMetrics. InfoMetrics are processed by processInfoMetrics. 
If the familiy is one of the InfoMetrics then bail here. It will be processed later.